File: quicksort.cpp
#include <bits/stdc++.h>

void swap(int* first, int* second){
  int temp = *first;
  *first = *second;
  *second = temp;
}

int partition(int arr[], int left, int right, int pivot){
  while(left<=right){
    while(arr[left]<pivot)
      left++;
    while(arr[right]>pivot)
      right--;

    if(left<=right){
      swap(&arr[left], &arr[right]);
      left++;
      right--;
    }
  }
  return left;
}

/* Driver program to test above functions */
void quickSort(int arr[], int left, int right){
  if(left>=right) return;
  int pivot = arr[left + (right - left)/2];
  int index = partition(arr, left, right, pivot);
  quickSort(arr, left, index-1);
  quickSort(arr, index, right);
}

void printArray(int A[], int size) 
{ 
  int i; 
  for (i=0; i < size; i++) 
    printf("%d ", A[i]); 
  printf("\n"); 
} 

/* Driver program to test above functions */
int main() 
{ 
  int arr_size; 
  std::cin >> arr_size;
  int arr[arr_size]; 
  for (int i = 0; i < arr_size; ++i) {
    std::cin >> arr[i];
  }

  quickSort(arr, 0, arr_size - 1); 

  printArray(arr, arr_size); 
  return 0; 
} 

File: coursera/algorithms/QuickSort.java
public class QuickSort {
  public QuickSort(int[] arr) {
    this.sort(arr, 0, arr.length-1);
  } 

  public void swap(int arr[], int first, int second){
    int tmp = arr[first];
    arr[first] = arr[second];
    arr[second] = tmp;
  }

  public int partition(int[] arr, int left, int right) {
    int i = left, j = left, pivot = arr[right];
    while (j < right) {
      if (arr[j] < pivot) {
        swap(arr, j, i);
        i++;
      }
      j++;
    }
    swap(arr, i, right);
    return i;
  }


  public void sort(int[] arr, int left, int right) {
    if (left < right) {
      int partition = partition(arr, left, right);
      this.sort(arr, left, partition-1);
      this.sort(arr, partition+1, right);
    }
  }

  public static void main(String[] args) {
    int[] arr = {10,9,8};
    QuickSort quickSort = new QuickSort(arr);
    for (int i : arr) {
      System.out.println(i);
    }
  }
}
